### DISCUSSION TITLE
ROS2 - Enable data sharing/ zero copy

### DISCUSSION THREAD

**Discussion by Nir-Az** [eProsima/Fast-DDS#2550] [Q&A] [answered]
Hi,

In order to use data sharing and loan messages on ROS2 with FastDDS MW do I need to set some configuration parameters?
What is the recommended flow to configure ROS2 to use FastDDS IPC working with zero copy ?

Thanks

**Comment by EduPonz** [comment, marked as answer]
Hi @Nir-Az ,

1. Data sharing is enabled and configured as `OFF` by default in Fast DDS when using it through ROS 2, so to use it in ROS 2:
    1. [These constrains](https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/datasharing.html#constraints) must be met.
    2. You'd need to enable it using XML (see ros2/rmw_fastrtps#579).
4. As for Zero-copy, which is the combination of data sharing and sample loaning in both sender and receiver side, the ROS 2 subscriber uses loans whenever it's possible, i.e. whenever [these constrains](https://fast-dds.docs.eprosima.com/en/latest/fastdds/use_cases/zero_copy/zero_copy.html#constraints) are met. As for the publication side, ROS 2 does not provide documentation about the feature just yet, but you can see an example [here](https://github.com/ros2/demos/blob/master/demo_nodes_cpp/src/topics/talker_loaned_message.cpp).

Hope that helps!

**Comment by Nir-Az** [comment]
@EduPonz Helps a lot thanks,

1. I must use the `borrow_loaned_message()` API in order for zero copy to be used right? cause I will need to constructe my data direcly on the shared memory?
2. Do you know of an implementation that can transport sensor_msgs/Image topic at zero copy? I know it's size is dynamic allocated so it cannot be used and I will need a dedicated topic for the fixed image size?


4.

**Reply by EduPonz** [reply]
Hi @Nir-Az ,

We are not aware of such implementation at the moment. However, I think the ROS 2 community would very much appreciate such contribution (if it does not exist yet).

**Reply by Nir-Az** [reply]
Thanks

**Reply by Ryanf55** [reply]
Given that statement, is it actually true that the V4L2 package here is zero copy? 
https://gitlab.com/boldhearts/ros2_v4l2_camera/-/tree/rolling#features

**Reply by sirawats** [reply]
I think that V4L2 use [intra-process communication ](https://docs.ros.org/en/foxy/Tutorials/Demos/Intra-Process-Communication.html)which is zero copy by default
